File: go/stack/ArrayStack.go
package stack

import (
	"errors"
	"fmt"
)

//顺序栈的实现（基于数组）

const DefaultCapacity = 16

type ArrayStack struct {
	data     []interface{}
	size     int
	capacity int
}

//新建数组
func New(capacity int) *ArrayStack {
	return &ArrayStack{
		data:     make([]interface{}, 0, capacity),
		size:     0,
		capacity: capacity,
	}
}

func Default() *ArrayStack {
	return New(DefaultCapacity)
}

//压栈，往栈中加入数据
func (stack *ArrayStack) Push(val interface{}) error {
	if stack.size == stack.capacity {
		return errors.New("stack is full")
	}

	stack.data = append(stack.data, val)
	stack.size++
	return nil
}

//出栈，取出栈顶数据
func (stack *ArrayStack) Pop() (interface{}, error) {
	if stack.Empty() {
		return nil, errors.New("stack is empty")
	}

	stack.size--
	result := stack.data[stack.size]
	stack.data[stack.size] = nil

	return result, nil
}

//获取栈顶元素
func (stack *ArrayStack) Peek() (interface{}, error) {
	if stack.Empty() {
		return nil, errors.New("stack is empty")
	}
	return stack.data[stack.size-1], nil
}

func (stack *ArrayStack) Size() int {
	return stack.size
}

func (stack *ArrayStack) Empty() bool {
	return stack.size <= 0
}

func (stack *ArrayStack) PrintData() {
	if stack.Empty() {
		fmt.Println("stack is empty")
	} else {
		for i := stack.size - 1; i >= 0; i-- {
			fmt.Print(stack.data[i], " ")
		}
		fmt.Println()
	}
}

File: python/stack/array_stack.py

# 使用数组实现的栈

class ArrayStack:

    def __init__(self, capacity: int):
        self.data = []
        self.size = 0
        self.capacity = capacity

    # 入栈
    def push(self, val: int):
        if self.size >= self.capacity:
            raise RuntimeError('the stack is full')
        self.data.append(val)
        self.size += 1

    # 出栈
    def pop(self) -> int:
        if self.size <= 0:
            raise RuntimeError('the stack is empty')
        res = self.data[self.size - 1]
        self.data = self.data[:self.size - 1]
        self.size -= 1
        return res


if __name__ == '__main__':
    stack = ArrayStack(10)
    stack.push(1)
    stack.push(6)
    stack.push(3)

    print(stack.pop())
    print(stack.pop())
    print(stack.pop())

    stack.push(10)
    print(stack.pop())
